SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AsyncDemotion.swift

```diff
@@ -0,0 +1,326 @@
+//===--- AsyncDemotion.swift -----------------------------------------------==//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SIL
+
+enum Executor: CustomStringConvertible, Hashable {
+  case selfActor            // self argument of this function
+  case nonisolated          // the global, concurrent executor
+  case unknown(HashableValue)
+
+  var description: String {
+    switch self {
+      case .selfActor: return "self"
+      case .nonisolated: return "nonisolated"
+      case let .unknown(value): return "\(value)"
+    }
+  }
+
+  init(_ op: Operand) {
+    self = .init(op.value)
+  }
+
+  init(_ val: Value) {
+    var finder = FindExecutor()
+    self = finder.search(val)
+  }
+
+  // Finds the executor corresponding to a hop instruction's operand.
+  struct FindExecutor : ValueUseDefWalker {
+    private var def: Executor?
+    internal var walkUpCache = WalkerCache<SmallProjectionPath>()
+
+    mutating func search(_ val: Value) -> Executor {
+      // assert(val.type == $Optional<Builtin.Executor>)
+      _ = walkUp(value: val, path: SmallProjectionPath(.anyValueFields))
+      guard let result = def else {
+        fatalError("could not find def for value:\n \(val)")
+      }
+      return result
+    }
+
+    internal mutating func rootDef(value: Value,
+                                   path: SmallProjectionPath) -> WalkResult {
+      log("FindExecutor.rootDef -> \(value)")
+
+      if let arg = value as? FunctionArgument {
```

Comment:
You can use a switch here (see my other comment)

Reply:
this conditional block got more complex than a switch can handle I think